From a52f49e53140488a25cfd6ef3c0867033f1ff222 Mon Sep 17 00:00:00 2001
From: "Fan, Yugang" <yugang.fan@intel.com>
Date: Tue, 20 Aug 2019 20:52:43 +0800
Subject: [PATCH] Use BRGA as default format when persist.waydroid.use_bgr=true

Some hardware only support BGRA format, we need config up-level to support this when using software rendering.

Adapted from Fan Yugang's <yugang.fan@intel.com> commit for EFI color format (I1fb2ba042522faac160440e5b660c01a24db2489)

Change-Id: I5487984835a1930420102eb13ebe0af88bf8de82
---
 .../CompositionEngine/src/RenderSurface.cpp     | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/services/surfaceflinger/CompositionEngine/src/RenderSurface.cpp b/services/surfaceflinger/CompositionEngine/src/RenderSurface.cpp
index 3fcd9d155..f790cefc9 100644
--- a/services/surfaceflinger/CompositionEngine/src/RenderSurface.cpp
+++ b/services/surfaceflinger/CompositionEngine/src/RenderSurface.cpp
@@ -31,6 +31,7 @@
 #include <ui/GraphicBuffer.h>
 #include <ui/Rect.h>
 #include <utils/Trace.h>
+#include <cutils/properties.h>
 
 #include "DisplayHardware/HWComposer.h"
 
@@ -67,11 +68,23 @@ bool RenderSurface::isValid() const {
 
 void RenderSurface::initialize() {
     ANativeWindow* const window = mNativeWindow.get();
+    char value[PROPERTY_VALUE_MAX] = {};
+    int format = HAL_PIXEL_FORMAT_RGBA_8888;
 
     int status = native_window_api_connect(window, NATIVE_WINDOW_API_EGL);
     ALOGE_IF(status != NO_ERROR, "Unable to connect BQ producer: %d", status);
-    status = native_window_set_buffers_format(window, HAL_PIXEL_FORMAT_RGBA_8888);
-    ALOGE_IF(status != NO_ERROR, "Unable to set BQ format to RGBA888: %d", status);
+	
+	// Use persist.waydroid.use_bgr=true to swap blue/red color space
+    if (property_get("persist.waydroid.use_bgr", value, NULL) > 0){
+        ALOGI("Use BGRA_8888 as surface format when EGL is \"%s\"", value);
+        if (!strcmp("true", value)) {
+            format = HAL_PIXEL_FORMAT_BGRA_8888;
+        }
+    }
+
+    status = native_window_set_buffers_format(window, format);
+    ALOGE_IF(status != NO_ERROR, "Unable to set BQ format to %d: %d", format, status);
+
     status = native_window_set_usage(window, GRALLOC_USAGE_HW_RENDER);
     ALOGE_IF(status != NO_ERROR, "Unable to set BQ usage bits for GPU rendering: %d", status);
 }
-- 
2.25.1

